external-crates/move/crates/move-analyzer/src/completion.rs

```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
+) -> bool {
+    if info.addresses.iter().any(|(_, a)| a == &pkg_addr) {
+        return true;
+    }
+
+    let all_identifiers = symbols
+        .file_mods
+        .values()
+        .flatten()
+        .map(|mdef| mdef.ident)
+        .collect::<BTreeSet<_>>();
+    for mod_ident in all_identifiers.into_iter() {
+        match mod_ident.address {
+            Address::Numerical {
+                name: _,
+                value,
+                name_conflict: _,
+            } if value.value == pkg_addr => return true,
+            _ => (),
+        }
+    }
+    false
+}
+
+/// Check if a given name represents a package within the current program
+fn is_package_name(symbols: &Symbols, info: &AliasAutocompleteInfo, pkg_name: Name) -> bool {
+    if info.addresses.contains_key(&pkg_name.value) {
+        return true;
+    }
+
+    let all_identifiers = symbols
+        .file_mods
+        .values()
+        .flatten()
+        .map(|mdef| mdef.ident)
+        .collect::<BTreeSet<_>>();
+    for mod_ident in all_identifiers.into_iter() {
+        match mod_ident.address {
+            Address::NamedUnassigned(name) if name == pkg_name => return true,
+            Address::Numerical {
+                name,
+                value: _,
+                name_conflict: _,
+            } if name == Some(pkg_name) => return true,
+            _ => (),
+        }
+    }
+    false
+}
+
+/// Get all packages that could be a target of auto-completion, whether they are part of
+/// `AliasAutocompleteInfo` or not.
+fn all_packages(symbols: &Symbols, info: &AliasAutocompleteInfo) -> BTreeSet<String> {
+    let mut addresses = BTreeSet::new();
+    for (n, a) in &info.addresses {
+        addresses.insert(n.to_string());
+        addresses.insert(a.to_string());
+    }
+
+    let all_identifiers = symbols
+        .file_mods
+        .values()
+        .flatten()
+        .map(|mdef| mdef.ident)
+        .collect::<BTreeSet<_>>();
+    for mod_ident in all_identifiers.into_iter() {
+        match mod_ident.address {
+            Address::Numerical {
+                name,
+                value,
+                name_conflict: _,
+            } => {
+                if let Some(n) = name {
+                    addresses.insert(n.to_string());
+                }
+                addresses.insert(value.to_string());
+            }
+            Address::NamedUnassigned(n) => {
+                addresses.insert(n.to_string());
+            }
+        }
+    }
+
+    addresses
+}
+
+/// Handle path auto-completion at a given position. The gist of this approach is to first identify
+/// what the first component of the access chain represents (as it may be a package, module or a
+/// member) and if the chain has other components, recursively process them in turn to either
+/// - finish auto-completion if cursor is on a given component's identifier
+/// - identify what the subsequent component represents and keep going
+fn path_completions(
```

Comment:
```suggestion
/// Handle name chain auto-completion at a given position. The gist of this approach is to first identify
/// what the first component of the access chain represents (as it may be a package, module or a
/// member) and if the chain has other components, recursively process them in turn to either
/// - finish auto-completion if cursor is on a given component's identifier
/// - identify what the subsequent component represents and keep going
fn name_chain_completions(
```